app/controllers/cloud_subnet_controller.rb

```diff
@@ -12,4 +12,253 @@ class CloudSubnetController < ApplicationController
   def self.display_methods
     %w(instances cloud_subnets)
   end
+
+  def button
+    @edit = session[:edit] # Restore @edit for adv search box
+    params[:display] = @display if %w(vms instances images).include?(@display)
+    params[:page] = @current_page unless @current_page.nil? # Save current page for list refresh
+
+    @refresh_div = "main_div"
+    return tag("CloudSubnet") if params[:pressed] == "cloud_subnet_tag"
+    delete_subnets if params[:pressed] == 'cloud_subnet_delete'
+
+    if params[:pressed] == "cloud_subnet_edit"
+      checked_subnet_id = get_checked_subnet_id(params)
+      javascript_redirect :action => "edit", :id => checked_subnet_id
+    elsif params[:pressed] == "cloud_subnet_new"
+      javascript_redirect :action => "new"
+    elsif !flash_errors? && @refresh_div == "main_div" && @lastaction == "show_list"
+      replace_gtl_main_div
+    else
+      render_flash
+    end
+  end
+
+  def cancel_action(message)
+    session[:edit] = nil
+    @breadcrumbs.pop if @breadcrumbs
+    javascript_redirect :action    => @lastaction,
+                        :id        => @subnet.id,
+                        :display   => session[:cloud_subnet_display],
+                        :flash_msg => message
+  end
+
+  def cloud_subnet_form_fields
+    assert_privileges("cloud_subnet_edit")
+    subnet = find_by_id_filtered(CloudSubnet, params[:id])
+    render :json => {
+      :name       => subnet.name,
+      :cidr       => subnet.cidr,
+      :gateway    => subnet.gateway,
+      :ip_version => subnet.ip_version
+    }
+  end
+
+  def new
+    assert_privileges("cloud_subnet_new")
+    @subnet = CloudSubnet.new
+    @in_a_form = true
+    @network_provider_choices = {}
+    ExtManagementSystem.where(type: "ManageIQ::Providers::Openstack::NetworkManager").find_each { |ems| @network_provider_choices[ems.name] = ems.id }
+    # Todo(gildub): Replace with angular lookup to narrow choice dynamically
+    @network_choices = {}
+    CloudNetwork.all.each { |network| @network_choices[network.name] = network.ems_ref }
+    @cloud_tenant_choices = {}
+    CloudTenant.all.each { |tenant| @cloud_tenant_choices[tenant.name] = tenant.id }
```

Comment:
Seems like you're assuming tenants and networks have unique names, is that always so?